### PR TITLE
feat: capped minter v2 minter role

### PR DESCRIPTION
**Pull request by marcomariscal**
## Description

Add `AccessControl` to `ZkCappedMinterV2`

- Inherit from OpenZeppelin AccessControl
- Add minter role (separate from admin role)
- Admin can grant/revoke minter role but cannot mint by default
- Only addresses with minter role can mint tokens

Kept all testing functionality the same as V1, but updated according to these changes. Also, included one more test for `testFuzz_RevertIf_AdminMintsByDefault`

### CODE REVIEW

File: l2-contracts/src/ZkCappedMinterV2.sol [hunk 1]
```diff
@@ -1,39 +1,40 @@
...
   uint256 public minted = 0;
 
+  bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+
   /// @notice Error for when the cap is exceeded.
   error ZkCappedMinterV2__CapExceeded(address minter, uint256 amount);
 
-  /// @notice Error for when the caller is not the admin.
+  /// @notice Error for when the account is unauthorized.
   error ZkCappedMinterV2__Unauthorized(address account);
```
Comment (alexkeating) [inline l2-contracts/src/ZkCappedMinterV2.sol:27]: We can probably remove this error in favor of this one https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/IAccessControl.sol#L13

File: l2-contracts/test/ZkCappedMinterV2.t.sol [hunk 2]
```diff
@@ -42,14 +43,21 @@ contract Mint is ZkCappedMinterV2Test {
     _amount = bound(_amount, 1, MAX_MINT_SUPPLY);
     vm.assume(_cap > _amount);
     vm.assume(_receiver != address(0) && _receiver != initMintReceiver);
+    vm.assume(_minter != address(0));
+
     ZkCappedMinterV2 cappedMinter = createCappedMinter(_cappedMinterAdmin, _cap);
+
     vm.prank(_cappedMinterAdmin);
+    cappedMinter.grantRole(MINTER_ROLE, _minter);
```
Comment (alexkeating) [inline l2-contracts/test/ZkCappedMinterV2.t.sol:51]: May be good to put this in a helper

**Review by alexkeating — COMMENTED** [review]
Looks good some small comments

File: l2-contracts/src/ZkCappedMinterV2.sol [hunk 3]
```diff
@@ -46,9 +47,9 @@ contract ZkCappedMinterV2 {
     TOKEN.mint(_to, _amount);
   }
 
-  /// @notice Reverts if msg.sender is not the contract admin.
+  /// @notice Reverts if the account is unauthorized.
   function _revertIfUnauthorized() internal view {
-    if (msg.sender != ADMIN) {
+    if (!hasRole(MINTER_ROLE, msg.sender)) {
```
Comment (alexkeating) [inline l2-contracts/src/ZkCappedMinterV2.sol:52]: We might want to use this instead. https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/AccessControl.sol#L86

Comment (alexkeating) [reply l2-contracts/src/ZkCappedMinterV2.sol:52]: Same in the upstream prs

File: l2-contracts/test/ZkCappedMinterV2.t.sol (hunk 2 above)
Comment (marcomariscal) [reply l2-contracts/test/ZkCappedMinterV2.t.sol]: https://github.com/ScopeLift/zk-governance/pull/4/commits/6ee3fb120966b14f9b33f8224a926a88fc2dae21

File: l2-contracts/src/ZkCappedMinterV2.sol (hunk 3 above)
Comment (marcomariscal) [reply l2-contracts/src/ZkCappedMinterV2.sol]: https://github.com/ScopeLift/zk-governance/pull/4/commits/72ef93c12a0535d82f7ad289ae4abb14024ff755

also removed the `Unauthorized` error since we are using `_checkRole`; this version of oz uses string reverts, so the test checks accordingly.

will update in the subsequent pr's.

File: l2-contracts/src/ZkCappedMinterV2.sol (hunk 1 above)
Comment (marcomariscal) [reply l2-contracts/src/ZkCappedMinterV2.sol]: sounds good; see this reply for more context:

https://github.com/ScopeLift/zk-governance/pull/4#discussion_r1876994336

File: l2-contracts/foundry.toml [hunk 4]
```diff
@@ -8,10 +9,10 @@
     "@openzeppelin/contracts-upgradeable/=lib/openzeppelin-contracts-upgradeable/contracts/",
     "@openzeppelin/foundry-upgrades/=lib/openzeppelin-foundry-upgrades/src/",
     "@murky/=lib/murky/",
+    "src/=src/",
```
Comment (alexkeating) [inline l2-contracts/foundry.toml:12]: Is this necessary?

File: l2-contracts/foundry.toml [hunk 5]
```diff
@@ -1,5 +1,6 @@
 [profile.default]
   evm_version = "paris"
+  fs_permissions = [{ access = "read", path = "./zkout" }]
```
Comment (alexkeating) [inline l2-contracts/foundry.toml:3]: Also, why was this added?

File: l2-contracts/package.json [hunk 6]
```diff
@@ -3,10 +3,10 @@
   "license": "MIT",
   "scripts": {
     "clean": "forge clean && npx hardhat clean",
-    "compile": "forge build && npx hardhat compile",
+    "compile": "forge build --zksync && npx hardhat compile",
     "local-node": "npx hardhat node-zksync",
     "script": "node script/RunScript.js",
-    "foundry-test": "forge test --no-match-path .integration.t.sol --zksync",
+    "foundry-test": "forge test --match-path 'test/*.t.sol' --no-match-path 'test/*.integration.t.sol' --zksync",
```
Comment (alexkeating) [inline l2-contracts/package.json:9]: What is being excluded with this match-path? Is it necessary?

**Review by alexkeating — APPROVED** [review]
One comment looks good
